galley/pkg/crd/synchronizer.go

```diff
@@ -0,0 +1,364 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package crd
+
+import (
+	errs "errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/workqueue"
+
+	"istio.io/istio/galley/pkg/change"
+	"istio.io/istio/pkg/log"
+
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Synchronizer is a Kubernetes controller that keeps a synchronized copy of an source set of CRDs in
+// destination APIGroup/Version(s). The mapping from source to destination is specified by a mapping object.
+// A passed-in listener can be used to detect when target CRDs are added/deleted.
+type Synchronizer struct {
+
+	// mutex for controlling the start/stop behavior.
+	stateLock sync.Mutex
```

Comment:
It's not clear what field(s) `stateLock` is guarding. Convention is usually to nest the mutex with the protected vars, e.g.

```golang
    stateLock sync.Mutex
    foo       string
    bar       string

    <... some other non-guarded fields...>
```